src/native/marker-index.cc

```diff
@@ -0,0 +1,420 @@
+
+  rotation_pivot->left_extent = rotation_root->left_extent.Traverse(rotation_pivot->left_extent);
+
+  rotation_pivot->right_marker_ids.insert(rotation_root->right_marker_ids.begin(), rotation_root->right_marker_ids.end());
+
+  for (auto it = rotation_pivot->left_marker_ids.begin(); it != rotation_pivot->left_marker_ids.end();) {
+    if (rotation_root->left_marker_ids.count(*it)) {
+      rotation_root->left_marker_ids.erase(*it);
+      ++it;
+    } else {
+      rotation_root->right_marker_ids.insert(*it);
+      rotation_pivot->left_marker_ids.erase(it++);
+    }
+  }
+}
+
+void MarkerIndex::RotateNodeRight(Node *rotation_pivot) {
+  Node *rotation_root = rotation_pivot->parent;
+
+  if (rotation_root->parent) {
+    if (rotation_root->parent->left == rotation_root) {
+      rotation_root->parent->left = rotation_pivot;
+    } else {
+      rotation_root->parent->right = rotation_pivot;
+    }
+  } else {
+    root = rotation_pivot;
+  }
+  rotation_pivot->parent = rotation_root->parent;
+
+  rotation_root->left = rotation_pivot->right;
+  if (rotation_root->left) {
+    rotation_root->left->parent = rotation_root;
+  }
+
+  rotation_pivot->right = rotation_root;
+  rotation_root->parent = rotation_pivot;
+
+  rotation_root->left_extent = rotation_root->left_extent.Traversal(rotation_pivot->left_extent);
+
+  for (auto it = rotation_root->left_marker_ids.begin(); it != rotation_root->left_marker_ids.end(); ++it) {
+    if (!rotation_pivot->start_marker_ids.count(*it)) {
+      rotation_pivot->left_marker_ids.insert(*it);
+    }
+  }
+
+  for (auto it = rotation_pivot->right_marker_ids.begin(); it != rotation_pivot->right_marker_ids.end();) {
+    if (rotation_root->right_marker_ids.count(*it)) {
+      rotation_root->right_marker_ids.erase(*it);
+      ++it;
+    } else {
+      rotation_root->left_marker_ids.insert(*it);
+      rotation_pivot->right_marker_ids.erase(it++);
+    }
+  }
+}
+
+void MarkerIndex::GetStartingAndEndingMarkersWithinSubtree(const Node *node, unordered_set<MarkerId> *starting, unordered_set<MarkerId> *ending) {
+  starting->insert(node->start_marker_ids.begin(), node->start_marker_ids.end());
+  ending->insert(node->end_marker_ids.begin(), node->end_marker_ids.end());
+  if (node->left) {
+    GetStartingAndEndingMarkersWithinSubtree(node->left, starting, ending);
+  }
+  if (node->right) {
+    GetStartingAndEndingMarkersWithinSubtree(node->right, starting, ending);
+  }
+}
+
+void MarkerIndex::PopulateSpliceInvalidationSets(SpliceResult *invalidated, const Node *start_node, const Node *end_node, const unordered_set<MarkerId> &starting_inside_splice, const unordered_set<MarkerId> &ending_inside_splice, bool is_insertion) {
+  invalidated->touch.insert(start_node->end_marker_ids.begin(), start_node->end_marker_ids.end());
+  invalidated->touch.insert(end_node->start_marker_ids.begin(), end_node->start_marker_ids.end());
+
+  for (const MarkerId &id : start_node->right_marker_ids) {
```

Comment:
You might just want to do `for (MarkerId id : start_node->right_marker_ids)`, since marker ids are primitives now.